tree_backfiller/src/tree.rs

```diff
@@ -0,0 +1,202 @@
+};
+use std::str::FromStr;
+use std::sync::Arc;
+use thiserror::Error as ThisError;
+use tokio::sync::mpsc::Sender;
+
+const GET_SIGNATURES_FOR_ADDRESS_LIMIT: usize = 1000;
+
+#[derive(Debug, Clone, Default, Args)]
+pub struct ConfigBackfiller {
+    /// Solana RPC URL
+    #[arg(long, env)]
+    pub solana_rpc_url: String,
+}
+
+#[derive(ThisError, Debug)]
+pub enum TreeErrorKind {
+    #[error("solana rpc")]
+    Rpc(#[from] solana_client::client_error::ClientError),
+    #[error("anchor")]
+    Achor(#[from] anchor_client::anchor_lang::error::Error),
+    #[error("perkle serialize")]
+    PerkleSerialize(#[from] plerkle_serialization::error::PlerkleSerializationError),
+    #[error("perkle messenger")]
+    PlerkleMessenger(#[from] plerkle_messenger::MessengerError),
+    #[error("queue send")]
+    QueueSend(#[from] tokio::sync::mpsc::error::SendError<Vec<u8>>),
+}
+#[derive(Debug, Clone)]
+pub struct TreeHeaderResponse {
+    pub max_depth: u32,
+    pub max_buffer_size: u32,
+    pub creation_slot: u64,
+    pub size: usize,
+}
+
+impl TryFrom<ConcurrentMerkleTreeHeader> for TreeHeaderResponse {
+    type Error = TreeErrorKind;
+
+    fn try_from(payload: ConcurrentMerkleTreeHeader) -> Result<Self, Self::Error> {
+        let size = merkle_tree_get_size(&payload)?;
+        Ok(Self {
+            max_depth: payload.get_max_depth(),
+            max_buffer_size: payload.get_max_buffer_size(),
+            creation_slot: payload.get_creation_slot(),
+            size,
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct TreeResponse {
+    pub pubkey: Pubkey,
+    pub tree_header: TreeHeaderResponse,
+}
+
+impl TreeResponse {
+    pub fn from_rpc(pubkey: Pubkey, account: Account) -> Result<Self> {
+        let (header_bytes, _rest) = account.data.split_at(CONCURRENT_MERKLE_TREE_HEADER_SIZE_V1);
+        let header: ConcurrentMerkleTreeHeader =
+            ConcurrentMerkleTreeHeader::try_from_slice(header_bytes)?;
+
+        let (auth, _) = Pubkey::find_program_address(&[pubkey.as_ref()], &mpl_bubblegum::ID);
+
+        header.assert_valid_authority(&auth)?;
+
+        let tree_header = header.try_into()?;
+
+        Ok(Self {
+            pubkey,
+            tree_header,
+        })
+    }
+}
+
+pub async fn all(client: &Arc<RpcClient>) -> Result<Vec<TreeResponse>, TreeErrorKind> {
+    let config = RpcProgramAccountsConfig {
+        filters: Some(vec![RpcFilterType::Memcmp(Memcmp::new_raw_bytes(
+            0,
+            vec![1u8],
+        ))]),
+        account_config: RpcAccountInfoConfig {
+            encoding: Some(UiAccountEncoding::Base64),
+            ..RpcAccountInfoConfig::default()
+        },
+        ..RpcProgramAccountsConfig::default()
+    };
+
+    Ok(client
+        .get_program_accounts_with_config(&id(), config)
+        .await?
+        .into_iter()
+        .filter_map(|(pubkey, account)| TreeResponse::from_rpc(pubkey, account).ok())
+        .collect())
+}
+
+pub async fn crawl(
```

Comment:
I know this is all still WIP and I haven't fully studied it all, but at first glance seems like `crawl` could be a method of `TreeResponse`?

Reply:
I can. I had that thought yesterday as well.